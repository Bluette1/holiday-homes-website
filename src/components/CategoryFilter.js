import { connect } from 'react-redux';
import React, { useEffect } from 'react';
import { useStateIfMounted } from 'use-state-if-mounted';
import PropTypes from 'prop-types';
import axios from 'axios';
import { changeFilter, registerCategories } from '../actions/index';
import { httpProtocol, host, port } from '../envVariables';

const CategoryFilter = ({ handleFilterChange, registerCategories }) => {
  const [data, setData] = useStateIfMounted([]);
  useEffect(() => {
    axios.get(`${httpProtocol}://${host}:${port}/categories`)
      .then(response => {
        const categoryList = response.data.categories;
        registerCategories(categoryList);
        setData(categoryList);
      });
  }, []);

  const holidayHomeCategories = ['All', ...data];
  return (
    <>
      <select
        className="category-filter"
        data-testid="category-filter"
        onChange={e => handleFilterChange(e.target.value)}
      >
        <option>CATEGORIES</option>
        {holidayHomeCategories.map(option => (
          <option key={option} value={option}>
            {option}
          </option>
        ))}
      </select>
    </>
  );
};

CategoryFilter.propTypes = {
  handleFilterChange: PropTypes.func.isRequired,
  registerCategories: PropTypes.func.isRequired,
};

export default connect(
  null, { handleFilterChange: changeFilter, registerCategories },
)(CategoryFilter);
