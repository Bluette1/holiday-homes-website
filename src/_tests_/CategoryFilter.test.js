import React from 'react';
import {
  render, screen, cleanup, waitFor,
} from '@testing-library/react';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import '@testing-library/jest-dom';
import axios from 'axios';
import CategoryFilter from '../components/CategoryFilter';
import configureStore, { history } from '../store';
import { httpProtocol, host, port } from '../envVariables';

jest.mock('axios');

afterEach(cleanup);

const store = configureStore({
  user: { username: 'MaryS', authentication_token: 'token' },
});

const CategoryFilterWithStore = () => (
  <Provider store={store}>
    <React.StrictMode>
      <ConnectedRouter history={history}><CategoryFilter /></ConnectedRouter>
    </React.StrictMode>
  </Provider>
);
const categories = [
  'Cottage',
  'Mansion',
  'Lodge',
  'Country home',
  'Chalet',
  'Log cabin',
];
test('renders the categoryFilter', async () => {
  axios.get.mockImplementation(url => {
    switch (url) {
      case `${httpProtocol}://${host}:${port}/categories`:
        return Promise.resolve({ data: { categories } });
      default:
        return Promise.resolve({ data: [] });
    }
  });

  render(<CategoryFilterWithStore />);
  await waitFor(() => {
    const categoryFilter = document.querySelector('.category-filter');
    expect(categoryFilter).toBeTruthy();

    const selectEl = screen.getByTestId('category-filter');
    expect(categoryFilter).toContainElement(selectEl);
    expect(categoryFilter).toContainHTML('<option>CATEGORIES</option>');

    expect(selectEl).toContainHTML('<option>CATEGORIES</option>');
    expect(selectEl).toContainHTML('<option value="All">All</option>');
    expect(selectEl).toContainHTML('<option value="Cottage">Cottage</option>');
    expect(selectEl).toContainHTML('<option value="Mansion">Mansion</option>');
    expect(selectEl).toContainHTML('<option value="Country home">Country home</option>');
    expect(selectEl).toContainHTML('<option value="Chalet">Chalet</option>');
    expect(selectEl).toContainHTML('<option value="Log cabin">Log cabin</option>');
    expect(categoryFilter).toMatchSnapshot();
  });
});
