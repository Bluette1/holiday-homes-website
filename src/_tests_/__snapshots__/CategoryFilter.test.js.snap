// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the app 1`] = `
<select
  class="category-filter"
  data-testid="category-filter"
>
  <option>
    CATEGORIES
  </option>
  <option
    value="All"
  >
    All
  </option>
  <option
    value="Cottage"
  >
    Cottage
  </option>
  <option
    value="Mansion"
  >
    Mansion
  </option>
  <option
    value="Lodge"
  >
    Lodge
  </option>
  <option
    value="Country home"
  >
    Country home
  </option>
  <option
    value="Chalet"
  >
    Chalet
  </option>
  <option
    value="Log cabin"
  >
    Log cabin
  </option>
</select>
`;
